openlibrary/plugins/importapi/tests/test_code.py

```diff
@@ -38,13 +38,14 @@ def test_get_ia_record(monkeypatch, mock_site, add_languages) -> None:  # noqa F
     expected_result = {
         "authors": [{"name": "Drury, Bob"}],
         "description": ["The story of the great Ogala Sioux chief Red Cloud"],
-        "isbn": ["9781451654684", "1451654685"],
+        "isbn_10": ["1451654685"],
+        "isbn_13": ["9781451654684"],
         "languages": ["fre"],
         "lccn": ["2013003200"],
         "number_of_pages": 450,
         "oclc": "1226545401",
         "publish_date": "2013",
-        "publisher": "New York : Simon & Schuster",
+        "publishers": ["New York : Simon & Schuster"],
```

Comment:
We should probably handle splitting on ` : ` if it's present. "New York" should go in `publish_places: list[str]`.

Reply:
I made this change. I am directly modifying the list that is passed in; would it be better to copy that list to another value so as not to modify the input directly?

I also put this in a helper function on the theory this makes this logic easier to test, and also it's easier for me to follow what `get_ia_record()` is doing with the ancillary logic elsewhere, but that may just be a fact about me.

Reply:
extracting publisher and place from this field can be a bit tricky. Here's some code I have used to extract the (first) publisher from archive.org fields on another project, it shows some of the other punctuation (square brackes, semi-colons, and commas) which can be encountered. It's probably not perfect either.

```
def extract_publisher(loc_pub):
    if not loc_pub:
        return ''
    publisher = loc_pub
    sqb = re.search(r'\[.*\]', publisher)
    if sqb:
        publisher = publisher.replace(sqb.group(0), '')
    if ':' in publisher:
        p = re.search(r':(.*)$', publisher)
        publisher = p.group(1)
    if ';' in publisher:  # multiple publishers / distributors
        publisher = re.search(r'^([^;]*);', publisher).group(1)
    if ',' in publisher:
        p = re.search(r'^(.*,)+(.{%d,})$' % max((len(publisher) // 2 - 1), 6), publisher)
        if p:
            publisher = p.group(2)
    return publisher.strip()
```

Reply:
Thank you for pointing this out, @hornc. I spent probably too long trying to come up with a 'good' way to do this.

I've updated the extraction function to better handle cases with multiple publishers and locations. It is not perfect, and I think it is hard to read.

I tried to make it handle situations where there is one publisher and one or more locations, separated by a colon, or two or more publishers and locations, and those publishers are separated by a semi-colon. If something doesn't fit that pattern it should 'fall though' and everything in the Internet Archive metadata `publisher` value goes into `publishers: ["IA publisher metadata here"]` field.

However, the extraction does not succeed perfectly at this goal. I grabbed 1500 or so publishers that would be imported as part of this larger project, and ran the publisher extraction, and the publisher and location extraction against them. I have included the output here as a text file should anyone be curious. Comas as dividers, but also punctuation, and other 'non-standard' formatting really made this a challenge for me.

The format is:
- original input string from the IA metadata
- `"publisher only"`
- `(["location1", "location2" ...], ["publisher1", "publisher2", ...])`
[extracted_publishers.txt](https://github.com/internetarchive/openlibrary/files/10693233/extracted_publishers.txt)